File: packages/less/src/less/less_go/nested_at_rule.go
package less_go

// Note: Ruleset is now implemented in ruleset.go

// NestableAtRulePrototype represents the prototype functionality for nestable at-rules
type NestableAtRulePrototype struct {
	*Node
	Type     string
	Features any
	Rules    []any
}

// NewNestableAtRulePrototype creates a new NestableAtRulePrototype instance
func NewNestableAtRulePrototype() *NestableAtRulePrototype {
	return &NestableAtRulePrototype{
		Node: NewNode(),
	}
}

// IsRulesetLike returns true indicating this behaves like a ruleset
func (n *NestableAtRulePrototype) IsRulesetLike() bool {
	return true
}

// Accept visits the node with a visitor
func (n *NestableAtRulePrototype) Accept(visitor any) {
	if n.Features != nil {
		if v, ok := visitor.(interface{ Visit(any) any }); ok {
			n.Features = v.Visit(n.Features)
		}
	}
	if n.Rules != nil {
		if v, ok := visitor.(interface{ VisitArray([]any) []any }); ok {
			n.Rules = v.VisitArray(n.Rules)
		}
	}
}

// EvalTop evaluates the at-rule at the top level 
func (n *NestableAtRulePrototype) EvalTop(context any) any {
	var result any = n

	// Handle context
	ctx, ok := context.(map[string]any)
	if !ok {
		return result
	}

	// Render all dependent Media blocks
	mediaBlocks, hasMediaBlocks := ctx["mediaBlocks"].([]any)
	if hasMediaBlocks && len(mediaBlocks) > 1 {
		// Create empty selectors
		selector, err := NewSelector(nil, nil, nil, n.GetIndex(), n.FileInfo(), nil)
		if err != nil {
			return result
		}
		emptySelectors, err := selector.CreateEmptySelectors()
		if err != nil {
			return result
		}

		// Create new Ruleset - convert selectors to []any
		selectors := make([]any, len(emptySelectors))
		for i, sel := range emptySelectors {
			selectors[i] = sel
		}
		ruleset := NewRuleset(selectors, mediaBlocks, false, n.VisibilityInfo())
		ruleset.MultiMedia = true // Set MultiMedia to true for multiple media blocks
		n.SetParent(ruleset.Node, n.Node)
		result = ruleset
	}

	// Delete mediaBlocks and mediaPath from context
	delete(ctx, "mediaBlocks")
	delete(ctx, "mediaPath")

	return result
}

// EvalNested evaluates the at-rule in a nested context
func (n *NestableAtRulePrototype) EvalNested(context any) any {
	ctx, ok := context.(map[string]any)
	if !ok {
		return n
	}

	mediaPath, hasMediaPath := ctx["mediaPath"].([]any)
	if !hasMediaPath {
		mediaPath = []any{}
	}

	// Create path with current node
	path := append(mediaPath, n)

	// Extract the media-query conditions separated with `,` (OR)
	for i := 0; i < len(path); i++ {
		pathItem, ok := path[i].(*NestableAtRulePrototype)
		if !ok {
			continue
		}

		if pathItem.Type != n.Type {
			if mediaBlocks, hasMediaBlocks := ctx["mediaBlocks"].([]any); hasMediaBlocks && i < len(mediaBlocks) {
				// Remove element at index i
				ctx["mediaBlocks"] = append(mediaBlocks[:i], mediaBlocks[i+1:]...)
			}
			return n
		}

		var value any
		if valueNode, ok := pathItem.Features.(*Value); ok {
			value = valueNode.Value
		} else {
			value = pathItem.Features
		}

		// Convert to array if needed
		if arr, ok := value.([]any); ok {
			path[i] = arr
		} else {
			path[i] = []any{value}
		}
	}

	// Trace all permutations to generate the resulting media-query
	permuteResult := n.Permute(path)
	if permuteResult == nil {
		return n
	}

	permuteArray, ok := permuteResult.([]any)
	if !ok {
		return n
	}

	// Ensure every path is an array before mapping
	for _, p := range permuteArray {
		if _, ok := p.([]any); !ok {
			return n
		}
	}

	// Map paths to expressions
	expressions := make([]any, len(permuteArray))
	for idx, pathItem := range permuteArray {
		pathArray, ok := pathItem.([]any)
		if !ok {
			continue
		}

		// Convert fragments
		mappedPath := make([]any, len(pathArray))
		for i, fragment := range pathArray {
			if _, ok := fragment.(interface{ ToCSS(any) string }); ok {
				mappedPath[i] = fragment
			} else {
				mappedPath[i] = NewAnonymous(fragment, 0, nil, false, false, nil)
			}
		}

		// Insert 'and' between fragments
		for i := len(mappedPath) - 1; i > 0; i-- {
			andAnon := NewAnonymous("and", 0, nil, false, false, nil)
			mappedPath = append(mappedPath[:i], append([]any{andAnon}, mappedPath[i:]...)...)
		}

		expr, err := NewExpression(mappedPath, false)
		if err != nil {
			continue
		}
		expressions[idx] = expr
	}

	// Create new Value with expressions
	newValue, err := NewValue(expressions)
	if err == nil {
		n.Features = newValue
		n.SetParent(n.Features, n.Node)
	}

	// Return fake tree-node that doesn't output anything
	return NewRuleset([]any{}, []any{}, false, nil)
}

// Permute creates permutations of the given array
func (n *NestableAtRulePrototype) Permute(arr []any) any {
	if len(arr) == 0 {
		return []any{}
	} else if len(arr) == 1 {
		return arr[0]
	} else {
		result := []any{}
		rest := n.Permute(arr[1:])
		
		restArray, ok := rest.([]any)
		if !ok {
			return nil
		}

		firstArray, ok := arr[0].([]any)
		if !ok {
			return nil
		}

		for i := 0; i < len(restArray); i++ {
			restItem, ok := restArray[i].([]any)
			if !ok {
				restItem = []any{restArray[i]}
			}
			
			for j := 0; j < len(firstArray); j++ {
				combined := append([]any{firstArray[j]}, restItem...)
				result = append(result, combined)
			}
		}
		return result
	}
}

// BubbleSelectors bubbles selectors up the tree
func (n *NestableAtRulePrototype) BubbleSelectors(selectors []*Selector) {
	if selectors == nil {
		return
	}
	if len(n.Rules) == 0 {
		return
	}

	copiedSelectors := make([]*Selector, len(selectors))
	copy(copiedSelectors, selectors)

	// Convert selectors to []any
	anySelectors := make([]any, len(copiedSelectors))
	for i, sel := range copiedSelectors {
		anySelectors[i] = sel
	}
	newRuleset := NewRuleset(anySelectors, []any{n.Rules[0]}, false, nil)
	n.Rules = []any{newRuleset}
	n.SetParent(n.Rules, n.Node)
} 

File: packages/less/lib/less/tree/nested-at-rule.js
// @ts-check
/** @import { EvalContext, CSSOutput, TreeVisitor, FileInfo, VisibilityInfo } from './node.js' */
import Ruleset from './ruleset.js';
import Value from './value.js';
import Selector from './selector.js';
import Anonymous from './anonymous.js';
import Expression from './expression.js';
import * as utils from '../utils.js';
import Node from './node.js';

/**
 * @typedef {object} FunctionRegistry
 * @property {(name: string, func: Function) => void} add
 * @property {(functions: Object) => void} addMultiple
 * @property {(name: string) => Function} get
 * @property {() => Object} getLocalFunctions
 * @property {() => FunctionRegistry} inherit
 * @property {(base: FunctionRegistry) => FunctionRegistry} create
 */

/**
 * @typedef {Node & {
 *   features: Value,
 *   rules: Ruleset[],
 *   type: string,
 *   functionRegistry?: FunctionRegistry,
 *   multiMedia?: boolean,
 *   debugInfo?: { lineNumber: number, fileName: string },
 *   allowRoot?: boolean,
 *   _evaluated?: boolean,
 *   evalFunction: () => void,
 *   evalTop: (context: EvalContext) => Node | Ruleset,
 *   evalNested: (context: EvalContext) => Node | Ruleset,
 *   permute: (arr: Node[][]) => Node[][],
 *   bubbleSelectors: (selectors: Selector[] | undefined) => void,
 *   outputRuleset: (context: EvalContext, output: CSSOutput, rules: Node[]) => void
 * }} NestableAtRuleThis
 */

// A media query may only carry a media type at its front, before any
// conditions (https://drafts.csswg.org/mediaqueries-5/#typedef-media-query-list).
// Operators that join media-query parts, so a fragment leading with one is not
// a media type. Any other bare identifier is treated as a type, since unknown
// media types are valid non-matches in CSS, not syntax errors.
const MEDIA_QUERY_OPERATORS = ['and', 'or'];

/**
 * Whether a flattened media-query fragment leads with a media type, e.g.
 * `screen`, `only screen`, `print and (color)` or an unknown type like `foo`.
 * @param {Node & { value?: * }} fragment
 */
function startsWithMediaType(fragment) {
    let head;
    if (fragment.type === 'Keyword' || fragment.type === 'Anonymous') {
        head = fragment.value;
    } else if (fragment.type === 'Expression' && Array.isArray(fragment.value)) {
        const parts = fragment.value.filter(p => p && p.value !== undefined);
        let idx = 0;
        const first = parts[idx] && String(parts[idx].value).toLowerCase();
        if (first === 'not' || first === 'only') { idx++; }
        head = parts[idx] && parts[idx].value;
    }
    if (typeof head !== 'string' || head === '') {
        return false;
    }
    // Inspect only the first token. A media type is a bare identifier; a feature
    // condition begins with '(' - e.g. an escaped ~"(max-width: 1px)" is an
    // Anonymous whose whole value is "(max-width: 1px)", which is not a media type.
    const firstToken = head.trim().split(/[\s(]/)[0];
    return firstToken !== ''
        && MEDIA_QUERY_OPERATORS.indexOf(firstToken.toLowerCase()) < 0;
}

const NestableAtRulePrototype = {

    isRulesetLike() {
        return true;
    },

    /** @param {TreeVisitor} visitor */
    accept(visitor) {
        /** @type {NestableAtRuleThis} */
        const self = /** @type {NestableAtRuleThis} */ (/** @type {unknown} */ (this));
        if (self.features) {
            self.features = /** @type {Value} */ (visitor.visit(self.features));
        }
        if (self.rules) {
            self.rules = /** @type {Ruleset[]} */ (visitor.visitArray(self.rules));
        }
    },

    evalFunction: function () {
        /** @type {NestableAtRuleThis} */
        const self = /** @type {NestableAtRuleThis} */ (/** @type {unknown} */ (this));
        if (!self.features || !Array.isArray(self.features.value) || self.features.value.length < 1) {
            return;
        }

        const exprValues = /** @type {Node[]} */ (self.features.value);
        /** @type {Node | undefined} */
        let expr;
        /** @type {Node | undefined} */
        let paren;

        for (let index = 0; index < exprValues.length; ++index) {
            expr = exprValues[index];

            if ((expr.type === 'Keyword' || expr.type === 'Variable')
                && index + 1 < exprValues.length
                && (/** @type {Node & { noSpacing?: boolean }} */ (expr).noSpacing || /** @type {Node & { noSpacing?: boolean }} */ (expr).noSpacing == null)) {
                paren =  exprValues[index + 1];

                if (paren.type ===  'Paren' && /** @type {Node & { noSpacing?: boolean }} */ (paren).noSpacing) {
                    exprValues[index]= new Expression([expr, paren]);
                    exprValues.splice(index + 1, 1);
                    /** @type {Node & { noSpacing?: boolean }} */ (exprValues[index]).noSpacing = true;
                }
            }
        }
    },

    /** @param {EvalContext} context */
    evalTop(context) {
        /** @type {NestableAtRuleThis} */
        const self = /** @type {NestableAtRuleThis} */ (/** @type {unknown} */ (this));
        self.evalFunction();

        /** @type {Node | Ruleset} */
        let result = self;

        // Render all dependent Media blocks.
        if (context.mediaBlocks.length > 1) {
            const selectors = (new Selector([], null, null, self.getIndex(), self.fileInfo())).createEmptySelectors();
            result = new Ruleset(selectors, context.mediaBlocks);
            /** @type {Ruleset & { multiMedia?: boolean }} */ (result).multiMedia = true;
            result.copyVisibilityInfo(self.visibilityInfo());
            self.setParent(result, self);
        }

        delete context.mediaBlocks;
        delete context.mediaPath;

        return result;
    },

    /** @param {EvalContext} context */
    evalNested(context) {
        /** @type {NestableAtRuleThis} */
        const self = /** @type {NestableAtRuleThis} */ (/** @type {unknown} */ (this));
        self.evalFunction();

        let i;
        /** @type {Node | Node[]} */
        let value;
        const path = context.mediaPath.concat([self]);

        // Extract the media-query conditions separated with `,` (OR).
        for (i = 0; i < path.length; i++) {
            if (path[i].type !== self.type) {
                const blockIndex = context.mediaBlocks.indexOf(self);
                if (blockIndex > -1) {
                    context.mediaBlocks.splice(blockIndex, 1);
                }
                return self;
            }

            value = /** @type {NestableAtRuleThis} */ (path[i]).features instanceof Value ?
                /** @type {Node[]} */ (/** @type {NestableAtRuleThis} */ (path[i]).features.value) : /** @type {NestableAtRuleThis} */ (path[i]).features;
            path[i] = /** @type {Node} */ (/** @type {unknown} */ (Array.isArray(value) ? value : [value]));
        }

        // Trace all permutations to generate the resulting media-query.
        //
        // (a, b and c) with nested (d, e) ->
        //    a and d
        //    a and e
        //    b and c and d
        //    b and c and e
        self.features = new Value(self.permute(/** @type {Node[][]} */ (/** @type {unknown} */ (path))).map(
            /** @param {Node | Node[]} path */
            path => {
                path = /** @type {Node[]} */ (path).map(
                /** @param {Node & { toCSS?: Function }} fragment */
                    fragment => fragment.toCSS ? fragment : new Anonymous(/** @type {string} */ (/** @type {unknown} */ (fragment))));

                // A media type nested inside conditions must move ahead of them
                // so the flattened query stays valid (issue #3694, #3764).
                const types = /** @type {Node[]} */ (path).filter(startsWithMediaType);
                if (types.length && types.length < /** @type {Node[]} */ (path).length) {
                    path = types.concat(/** @type {Node[]} */ (path).filter(f => !startsWithMediaType(f)));
                }

                for (i = /** @type {Node[]} */ (path).length - 1; i > 0; i--) {
                /** @type {Node[]} */ (path).splice(i, 0, new Anonymous('and'));
                }

                return new Expression(/** @type {Node[]} */ (path));
            }));
        self.setParent(self.features, self);

        // Fake a tree-node that doesn't output anything.
        return new Ruleset([], []);
    },

    /**
     * @param {Node[][]} arr
     * @returns {Node[][]}
     */
    permute(arr) {
        if (arr.length === 0) {
            return [];
        } else if (arr.length === 1) {
            return /** @type {Node[][]} */ (/** @type {unknown} */ (arr[0]));
        } else {
            /** @type {Node[][]} */
            const result = [];
            const rest = this.permute(arr.slice(1));
            for (let i = 0; i < rest.length; i++) {
                for (let j = 0; j < arr[0].length; j++) {
                    result.push([arr[0][j]].concat(rest[i]));
                }
            }
            return result;
        }
    },

    /** @param {Selector[] | undefined} selectors */
    bubbleSelectors(selectors) {
        /** @type {NestableAtRuleThis} */
        const self = /** @type {NestableAtRuleThis} */ (/** @type {unknown} */ (this));
        if (!selectors) {
            return;
        }
        self.rules = [new Ruleset(utils.copyArray(selectors), [self.rules[0]])];
        self.setParent(self.rules, self);
    }
};

export default NestableAtRulePrototype;
